### PR TITLE
Add postal code validation

### LINKED ISSUE #6: Validate postal code
- [ ] The postal code should be either numbers and/or letters
- [ ] Countries define the postal code matching expression. Example: `####-####`

Check
- https://en.wikipedia.org/wiki/Postal_code

### PR DESCRIPTION
**Pull request by ttoomes**
This closes #6

### CODE REVIEW

**Comment by yordis** [comment]
Hello, @ttoomes! This is your first Pull Request that will be reviewed by [Ebert](https://ebertapp.io/), an automatic Code Review service. It will leave comments on this diff with potential issues and style violations found in the code as you push new commits. You can also see all the issues found on this Pull Request on its [review page](https://ebertapp.io/github/straw-hat-team/straw_hat_map/pulls/20). Please check our [documentation](https://docs.ebertapp.io) for more information.

File: lib/straw_hat_map/world/countries/country.ex [hunk 1]
```diff
@@ -7,6 +7,7 @@ defmodule StrawHat.Map.Country do
   use StrawHat.Map.Schema
   alias StrawHat.Map.Continents
   alias StrawHat.Map.State
+  alias StrawHat.Map.Regex
```
Comment (yordis) [inline lib/straw_hat_map/world/countries/country.ex]: Most of the time you are using the multi-alias/require/import/use syntax, but here you are using multiple single directives

File: lib/straw_hat_map/regex.ex [hunk 2]
```diff
@@ -0,0 +1,28 @@
+defmodule StrawHat.Map.Regex do
+  @behaviour Ecto.Type
+  def type, do: :string
+
+  @spec cast(Regex.t) :: {:ok, String.t} | :error
+  def cast(regex) do
+    case Regex.regex?(regex) do
+      true -> {:ok, Regex.source(regex)}
+      false -> :error
+    end
```
Comment (yordis) [inline lib/straw_hat_map/regex.ex]: There should be no trailing white-space at the end of a line.

File: lib/straw_hat_map/regex.ex [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  When dumping data to the database, we *expect* our casted type as a string
+    iex> Regex.dump("foo")
+    {:ok, "foo"}
+    iex> Regex.dump(1)
+    :error
+  """
+  @spec dump(String.t) :: {:ok, String.t} | :error
+  def dump(regex) when is_binary(regex), do: {:ok, regex}
+  def dump(_), do: :error
+end
```
Comment (yordis) [inline lib/straw_hat_map/regex.ex]: There should be a final \n at the end of each file.

File: lib/straw_hat_map/regex.ex [hunk 4]
```diff
@@ -0,0 +1,28 @@
+defmodule StrawHat.Map.Regex do
```
Comment (yordis) [inline lib/straw_hat_map/regex.ex]: Modules should have a @moduledoc tag.

File: lib/straw_hat_map/addresses/address.ex [hunk 5]
```diff
@@ -56,5 +56,27 @@ defmodule StrawHat.Map.Address do
...
+
+    changeset
+    |> validate_format(:postal_code, rule)
+  end
+
+  @spec get_postal_code_rule(Integer.t()) :: City.t() | nil | no_return
+  defp get_postal_code_rule(city) do
+    %City{state: %StrawHat.Map.State{country: %StrawHat.Map.Country{postal_code_rule: rule}}} =
+      StrawHat.Map.Repo.get!(City, city)
+      |> StrawHat.Map.Repo.preload([state: :country])
```
Comment (yordis) [inline lib/straw_hat_map/addresses/address.ex]: Use a function call when a pipeline is only one function long

File: lib/straw_hat_map/addresses/address.ex [hunk 6]
```diff
@@ -56,5 +56,27 @@ defmodule StrawHat.Map.Address do
...
-end
+
+  @spec validate_postal_code(Ecto.Changeset.t()) :: Ecto.Changeset.t()
+  defp validate_postal_code(changeset) do
+    rule =
+      get_field(changeset, :city_id)
+      |> get_postal_code_rule()
+
+    changeset
+    |> validate_format(:postal_code, rule)
```
Comment (yordis) [inline lib/straw_hat_map/addresses/address.ex]: Use a function call when a pipeline is only one function long

File: lib/straw_hat_map/addresses/address.ex [hunk 7]
```diff
@@ -56,5 +56,27 @@ defmodule StrawHat.Map.Address do
...
     |> assoc_constraint(:city)
+    |> validate_postal_code()
   end
-end
+
+  @spec validate_postal_code(Ecto.Changeset.t()) :: Ecto.Changeset.t()
+  defp validate_postal_code(changeset) do
+    rule =
+      get_field(changeset, :city_id)
+      |> get_postal_code_rule()
```
Comment (yordis) [inline lib/straw_hat_map/addresses/address.ex]: Use a function call when a pipeline is only one function long

File: lib/straw_hat_map/addresses/address.ex [hunk 8]
```diff
@@ -56,5 +56,27 @@ defmodule StrawHat.Map.Address do
...
+  defp get_postal_code_rule(city) do
+    %City{state: %StrawHat.Map.State{country: %StrawHat.Map.Country{postal_code_rule: rule}}} =
+      StrawHat.Map.Repo.get!(City, city)
+      |> StrawHat.Map.Repo.preload([state: :country])
+    case rule do
+      nil -> ~r/^\w+[ -]?\w+$/
+      _ -> rule
+    end
+  end
+end
```
Comment (yordis) [inline lib/straw_hat_map/addresses/address.ex]: There should be a final \n at the end of each file.

**Comment by yordis** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 9 possible new issues (including those that may have been commented here).
* 1 fixed issue! 🎉

But beware that this branch is **[12 commits behind the `straw-hat-team:master` branch](https://github.com/straw-hat-team/straw_hat_map/compare/ttoomes:master...straw-hat-team:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/straw-hat-team/straw_hat_map/pulls/20.

File: lib/straw_hat_map/world/countries/country.ex [hunk 9]
```diff
@@ -5,9 +5,12 @@ defmodule StrawHat.Map.Country do
...
   use StrawHat.Map.Schema
-  alias StrawHat.Map.Continents
-  alias StrawHat.Map.State
-
+  alias StrawHat.Map.{
+    Continents,
+    State,
+    Regex
+  }
+
```
Comment (yordis) [inline lib/straw_hat_map/world/countries/country.ex]: There should be no trailing white-space at the end of a line.

File: lib/straw_hat_map/addresses/address.ex [hunk 10]
```diff
@@ -57,5 +57,27 @@ defmodule StrawHat.Map.Address do
...
+      |> get_postal_code_rule()
+
+    validate_format(changeset, :postal_code, rule)
+  end
+
+  @spec get_postal_code_rule(Integer.t()) :: Regex.t()
+  defp get_postal_code_rule(city) do
+    %City{state: %StrawHat.Map.State{country: %StrawHat.Map.Country{postal_code_rule: rule}}} =
+      StrawHat.Map.Repo.get!(City, city)
+      |> StrawHat.Map.Repo.preload([state: :country])
```
Comment (yordis) [inline lib/straw_hat_map/addresses/address.ex]: Use a function call when a pipeline is only one function long

**Comment by yordis** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 2 possible new issues (including those that may have been commented here).
* 1 fixed issue! 🎉

You can see more details about this review at https://ebertapp.io/github/straw-hat-team/straw_hat_map/pulls/20.

**Comment by ttoomes** [comment]
@yordis Thank you for the feedback. This is my first foray into Elixir and contributing to an open-source project, so I appreciate your patience.

I'll work on making those changes this weekend.

**Comment by yordis** [comment]
@ttoomes no problem, we are here for help each other, thank you for the contribution.

Your PR is just perfect but because every single package that we are creating have that constraint we need to change it.

After giving it a thought, the best thing to do is to pass a 3rd parameter to the changeset that has the postal code regex in it.

```
def changeset(address, address_attrs, opts) do
```

With the required key for validate the address, pretty much move the look up to the interactor and pass down the keyword list. This way the Schema/changesets are agnostic of the database or anything like that, just a plain function and a struct.

**Comment by yordis** [comment]
@ttoomes any updates?

File: lib/straw_hat_map/world/countries/country.ex [hunk 11]
```diff
@@ -5,9 +5,9 @@ defmodule StrawHat.Map.Country do
   """
 
   use StrawHat.Map.Schema
-  alias StrawHat.Map.Continents
-  alias StrawHat.Map.State
-
+  alias StrawHat.Map.{Continents, State}
+  alias StrawHat.Map.Ecto.Regex
+
```
Comment (yordis) [inline lib/straw_hat_map/world/countries/country.ex]: There should be no trailing white-space at the end of a line.

File: lib/straw_hat_map/addresses/addresses.ex [hunk 12]
```diff
@@ -73,4 +77,15 @@ defmodule StrawHat.Map.Addresses do
     query = from(address in Address, where: address.id in ^address_ids)
     Repo.all(query)
   end
+
+  @spec get_postal_code_rule(Integer.t()) :: Regex.t()
+  defp get_postal_code_rule(city_id) do
+    %StrawHat.Map.City{state: %StrawHat.Map.State{country: %StrawHat.Map.Country{postal_code_rule: rule}}} =
+      StrawHat.Map.Repo.get!(City, city_id)
+      |> StrawHat.Map.Repo.preload([state: :country])
```
Comment (yordis) [inline lib/straw_hat_map/addresses/addresses.ex]: Use a function call when a pipeline is only one function long

**Comment by yordis** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 2 possible new issues (including those that may have been commented here).
* 1 fixed issue! 🎉

But beware that this branch is **[7 commits behind the `straw-hat-team:master` branch](https://github.com/straw-hat-team/straw_hat_map/compare/ttoomes:master...straw-hat-team:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/straw-hat-team/straw_hat_map/pulls/20.

**Comment by yordis** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 1 possible new issue (including those that may have been commented here).
* 1 fixed issue! 🎉

But beware that this branch is **[7 commits behind the `straw-hat-team:master` branch](https://github.com/straw-hat-team/straw_hat_map/compare/ttoomes:master...straw-hat-team:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/straw-hat-team/straw_hat_map/pulls/20.

**Comment by yordis** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 1 fixed issue! 🎉

But beware that this branch is **[7 commits behind the `straw-hat-team:master` branch](https://github.com/straw-hat-team/straw_hat_map/compare/ttoomes:master...straw-hat-team:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/straw-hat-team/straw_hat_map/pulls/20.

**Comment by yordis** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 1 fixed issue! 🎉

You can see more details about this review at https://ebertapp.io/github/straw-hat-team/straw_hat_map/pulls/20.

**Comment by ttoomes** [comment]
@yordis I was finally able to make the changes. I moved the postal code regex logic out of the address schema, and it's passed to the changeset as an option like you suggested.

**Comment by yordis** [comment]
@ttoomes Would you mind to make the coverage 100%? This will help you to learn about ExUnit and improve your skills and in the other side the package keep the minimal quality up.

Let me know if you wanna do it, but it is not a problem at all.

**Pull request merged**